cdap-ui/app/cdap/components/DataPrep/ColumnTextSelection/index.js

```diff
@@ -0,0 +1,277 @@
+          ['TR', 'TBODY'].indexOf(e.target.nodeName) === -1
+        ) {
+          if (this.props.onClose) {
+            this.props.onClose();
+          }
+        }
+      });
+    let highlightedHeader = document.getElementById('highlighted-header');
+    if (highlightedHeader) {
+      highlightedHeader.scrollIntoView();
+    }
+  }
+  componentDidUpdate() {
+    if (this.tetherRef) {
+      this.tetherRef.position();
+    }
+  }
+  componentWillUnmount() {
+    if (this.documentClick$) {
+      this.documentClick$.dispose();
+    }
+  }
+  applyDirective() {
+    this.onApply(this.state.textSelectionRange);
+  }
+  handleColNameChange(value, isChanged, keyCode) {
+    this.newColName = value;
+    if (keyCode === 13) {
+      this.applyDirective();
+    }
+  }
+  togglePopover() {
+    if (this.state.showPopover) {
+      this.setState({
+        showPopover: false,
+        textSelectionRange: {start: null, end: null, index: null},
+      });
+      this.newColName = null;
+      this.props.togglePopover(false);
+      return;
+    }
+  }
+  mouseDownHandler() {
+    this.textSelection = true;
+    this.togglePopover();
+  }
+  mouseUpHandler(head, index) {
+
+    let currentSelection = window.getSelection().toString();
+    let startRange, endRange;
+
+    if (this.textSelection && currentSelection.length) {
+      startRange = window.getSelection().getRangeAt(0).startOffset;
+      endRange = window.getSelection().getRangeAt(0).endOffset;
+      this.textSelection = false;
+      let textSelectionRange = {
+        start: startRange,
+        end: endRange,
+        index
+      };
+      this.setState({
+        showPopover: true,
+        textSelectionRange
+      });
+      this.props.onSelect({
+        textSelectionRange,
+        rowNumber: index
+      });
+      this.props.togglePopover(true);
+      this.newColName = this.props.columns[0] + '_copy';
+    } else {
+      if (this.state.showPopover) {
+        this.props.onSelect({textSelectionRange: null, showPopover:true, rowNumber: null});
+        this.togglePopover();
+      }
+    }
+  }
+  preventPropagation(e) {
+    e.stopPropagation();
+    e.nativeEvent.stopImmediatePropagation();
+    e.preventDefault();
+  }
+
+  render() {
+    let {data, headers} = DataPrepStore.getState().dataprep;
+    let column = this.props.columns[0];
+
+    const renderTableCell = (row, index, head, highlightColumn) => {
+      if (head !== highlightColumn) {
+        return (
+          <td
+            key={shortid.generate()}
+            className="gray-out"
+          >
+            <div className="gray-out">
+              {row[head]}
+            </div>
+          </td>
+        );
+      }
+      return (
+        <td
+          key={shortid.generate()}
+          className={CELLHIGHLIGHTCLASSNAME}
+          onMouseDown={this.mouseDownHandler}
+          onMouseUp={this.mouseUpHandler.bind(this, head, index)}
+        >
+          <div
+            className={CELLHIGHLIGHTCLASSNAME}
```

Comment:
Unfortunately this one I needed because when I click anywhere outside vs clicking on the popover I should be able to differentiate with the target. So any target with `CELLHIGHLIGHTCLASSNAME` won't trigger a close of column highlight mode. Ref: line 56.

I could instead have had one top level parent having a css class and detect any target if its a child of that parent but this looked simpler. Let me know if there is a better approach.